### PR TITLE
 add txpool interface && blocksync interface && consensus interface to lab-bcos

### PR DESCRIPTION
**Pull request by cyjseagull**
add txpool interface && blocksync interface && consensus interface to lab-bcos

### CODE REVIEW

File: libblocksync/Common.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * @brief : common functions and types of Blocksync modules
+ * @author: yujiechen
+ * @date: 2018-09-21
+ */
+#pragma once
+#include <libdevcore/Exceptions.h>
+#include <libp2p/Common.h>
+#include <set>
+using namespace dev::p2p;
```
Comment (wheatli) [inline libblocksync/Common.h]: Forbit to using namepace in header files

File: libblocksync/SyncInterface.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// for rpc && sdk: broad cast transaction to all nodes
+    virtual void broadCastTransactions() = 0;
+    /// for p2p: broad cast transaction to specified nodes
+    virtual void sendTransactions(NodeList const& _nodes) = 0;
+
+    /// abort sync and reset all status of blockSyncs
+    virtual void reset() = 0;
+    virtual bool forceSync() = 0;
+    static const int16_t c_ProtoclId = 2;
```
Comment (wheatli) [inline libblocksync/SyncInterface.h]: can't use static member due to multipul ledgers, we need supply a interface like setProtocolID

File: libconsensus/Common.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * @brief Common functions and types of Consensus module
+ * @author: yujiechen
+ * @date: 2018-09-21
+ */
+#pragma once
+namespace dev
+{
+namespace consensus
+{
+enum PBFTStatus
```
Comment (wheatli) [inline libconsensus/Common.h]: All about PBFT contents exist in derived class...

Comment (wheatli) [reply libconsensus/Common.h]: So maybe you can give a base class named ConsensusStatus, and let PBFTStatus derived from it

File: libconsensus/ConsensusInterface.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    /// stop the consensus module
+    virtual void stop() = 0;
+
+    /// get miner list
+    virtual h512s getMinerList() const = 0;
+    /// set the miner list
+    virtual void setMinerList(h512s const& _minerList) const = 0;
+
+    /// get status of consensus
+    virtual PBFTStatus pbftStatus() const = 0;
```
Comment (wheatli) [inline libconsensus/ConsensusInterface.h]: return ConsensusStatus const&, and derived class can return PBFTStatus const&

File: libtxpool/TxPoolInterface.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * @brief : interface of transaction pool
+ * @file : TxPoolInterface.h
+ * @author: yujiechen
+ * @date: 2018-09-21
+ */
+#pragma once
+#include <libethcore/Transaction.h>
+using namespace dev::eth;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: can not expand namespace in the header files

File: libtxpool/TxPoolInterface.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    /// push a transaction into the txpool
+    virtual void enqueue(bytesRef _data, h512 const& _nodeId) = 0;
+
+    /**
+     * @brief:mark transaction as future.
+     *        It wont be retured in topTransactions list until
+     *        a transaction with a preceeding nonce is imported or is dropped
+     * @param _t : Transaction hash
+     */
+    virtual void setFuture(h256 const& _t) = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: its not nessesary

File: libtxpool/TxPoolInterface.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @param _limit : _limit Max number of transactions to return.
+     * @param _avoid : Transactions to avoid returning.
+     * @return Transactions : up to _limit transactions
+     */
+    virtual Transactions topTransactions(
+        unsigned _limit, h256Hash const& _avoid = h256Hash()) const = 0;
+
+    /// get all transactions(maybe blocksync module need this interface)
+    virtual Transactions allTransactions() const = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: not essential api

File: libtxpool/TxPoolInterface.h [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+     *        It wont be retured in topTransactions list until
+     *        a transaction with a preceeding nonce is imported or is dropped
+     * @param _t : Transaction hash
+     */
+    virtual void setFuture(h256 const& _t) = 0;
+    /**
+     * @brief Remove transaction from the queue
+     * @param _txHash: transaction hash
+     */
+    virtual void drop(h256 const& _txHash) = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: we need one api named drop, another one named dropGood

Comment (wheatli) [reply libtxpool/TxPoolInterface.h]: Or you can merge them in an api, and give a param to indicate the difference

File: libtxpool/TxPoolInterface.h [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+     * @brief : submit a transaction through p2p, Verify and add transaction to the queue
+     * synchronously.
+     *
+     * @param _tx : Trasnaction data.
+     * @param _ik : Set to Retry to force re-addinga transaction that was previously dropped.
+     * @return ImportResult : Import result code.
+     */
+    virtual ImportResult import(Transaction const& _tx, IfDropped _ik = IfDropped::Ignore);
+    /// @returns the status of the transaction queue.
+    virtual TxPoolStatus status() const = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: I haven't got your aim...Plz check the design slice we have touched

File: libtxpool/TxPoolInterface.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        unsigned _limit, h256Hash const& _avoid = h256Hash()) const = 0;
+
+    /// get all transactions(maybe blocksync module need this interface)
+    virtual Transactions allTransactions() const = 0;
+    /**
+     * @brief : Get a hash set of transactions in the queue
+     *
+     * @return h256Hash : A hash set of all transactions in the queue
+     */
+    virtual h256Hash knownTransactions() const = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: where does it used?

File: libtxpool/TxPoolInterface.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    /// get current transaction num
+    virtual size_t currentTxNum() = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: pendingSize is a better name

File: libtxpool/TxPoolInterface.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * @brief submit a transaction through RPC/web3sdk
+     *
+     * @param _t : transaction
+     * @return std::pair<h256, Address> : maps from transaction hash to contract address
+     */
+    virtual std::pair<h256, Address> submitTransaction(Transaction const& _tx) = 0;
```
Comment (wheatli) [inline libtxpool/TxPoolInterface.h]: Plz make clear about the three methods abou enqueue, import and submitTx, what's the difference between them, how to use them... Plz check the design slice again

File: libtxpool/TxPoolInterface.h (hunk 7 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: has been renamed to pendingList

File: libblocksync/Common.h (hunk 1 above)
Comment (cyjseagull) [reply libblocksync/Common.h]: already removed, pls refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-3ac83c9b460313bfc6c61b73e9d0609bR26

File: libblocksync/SyncInterface.h (hunk 2 above)
Comment (cyjseagull) [reply libblocksync/SyncInterface.h]: already maintain `getProtocolID` interface, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-cc075a822bb837e4d4bc9fb325afef16R57

File: libconsensus/Common.h (hunk 3 above)
Comment (cyjseagull) [reply libconsensus/Common.h]: already modified to `ConsensusStatus`, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-9cb3c483050ad9fd820030d43705bde7R28

File: libconsensus/ConsensusInterface.h (hunk 4 above)
Comment (cyjseagull) [reply libconsensus/ConsensusInterface.h]: already modified to `virtual ConsensusStatus consensusStatus() const`, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-dd37ce1b0a5c8b129dd97d86be3da6bdR48

File: libtxpool/TxPoolInterface.h (hunk 5 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: the namespace using has been removed, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-ca5bfa40c4811cd40a43a8d6f65179d8R24

File: libtxpool/TxPoolInterface.h (hunk 6 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: already removed

File: libtxpool/TxPoolInterface.h (hunk 7 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-ca5bfa40c4811cd40a43a8d6f65179d8R54

File: libtxpool/TxPoolInterface.h (hunk 8 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: `dropGood ` has been added, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-ca5bfa40c4811cd40a43a8d6f65179d8R42

File: libtxpool/TxPoolInterface.h (hunk 10 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: already removed

File: libtxpool/TxPoolInterface.h (hunk 11 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: has been modified to `pendingSize`, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-ca5bfa40c4811cd40a43a8d6f65179d8R56

File: libtxpool/TxPoolInterface.h (hunk 12 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: has been modified to `submit` interface, if this interface is not appropriate, I'll modify it in the future.
please refer to:

https://github.com/FISCO-BCOS/lab-bcos/pull/69/files#diff-ca5bfa40c4811cd40a43a8d6f65179d8R63

File: libtxpool/TxPoolInterface.h (hunk 9 above)
Comment (cyjseagull) [reply libtxpool/TxPoolInterface.h]: this `import` interface refers to `libethereum/Interface.*`, I'll modify it in the future if it's not appropriate.

**Pull request merged**